### DISCUSSION TITLE
To SSR or not to SSR

### DISCUSSION THREAD

**Discussion by kennerstross** [nuxt/nuxt#17657] [Questions] [answered]
Does anyone have any guidance, thoughts, advice, best practices, etc. regarding whether one should have SSR = true or false? What are the considerations one way or another? What do most people do (well, I assume that most use SSR since that's the default)?

**Comment by benoitdemaegdt** [comment, marked as answer]
Hey @kennerstross  👋 

There is a pretty good article about it here (written by Nuxt3 team) 👉  https://v3.nuxtjs.org/guide/concepts/rendering
The "Pros" and "Cons" sections are very instructive and may guide you to the most appropriate choice for your use case.

I think that what must be understood, is that SSR is not always the best option. It depends on what you want to build.

Let's take an example.

If you want to build a personal website or a blog, then : 
- your content is static (the same content is displayed to all users)
- you will probably want your content to be indexed by Search Engine (this is a practice called SEO). One criteria is that your website should load as fast as possible.

SSR looks like a good option here because it will make your loading time (much) faster and easier to index.
However you might wonder if rendering the same content on the server side for each users is a good option. It will increase your infrastructure costs because you need to run a serverless function each and every time a user sees one of your page.
The best option here is probably SSG (Static Site Generation), [currently experimental in Nuxt3](https://github.com/nuxt/framework/discussions/515). All your pages will be rendered at build time (only once !) and then served to your users as HTML files. This option will enable your blog to load very fast, and will cost you near $0 (depending on your trafic) in infrastructure costs.

Let's take another example.
You are building a Saas. So you have a landing page (static content) then a signup form and your app.
- your landing page is very similar to having a blog. SSG or SSR is a good option here.
- A user needs to signup to be able to access your app/ Saas. This content is unique for each user so you cannot use SSG here.
Then you have 2 options here : SSR or SPA (client side rendering).
As stated in Nuxt3 documentation, you may choose SSR if you have a performance first approach, and SPA if you want more simplicity while coding your app. I would say one option is not better than the other, it depends on you and your team.

One last thing. If you go through Nuxt3 documentation you may find something called [Hybrid rendering](https://v3.nuxtjs.org/guide/concepts/rendering#hybrid-rendering). The idea behind this concept is to use SSG for static content, and SSR for dynamic content. I never used it before but it sounds promising ⭐

**Reply by kennerstross** [reply]
Thanks for the excellent answer, @benoitdemaegdt! I think the article you linked was added after I read most of the documentation, a while back. Anyway, for others, that article along with this Vue article I found - [https://vuejs.org/guide/scaling-up/ssr.html](https://vuejs.org/guide/scaling-up/ssr.html )- go a long way to demystifying all of this.

**Comment by riddla** [comment]
As a small agency focused on corporate sites we are using Static Site Generation, as it fits best for our use case. The sites of our clients consist mostly exclusive of static content.

For the small parts that are dynamic (like news, job openings) we do [exclude](https://nuxtjs.org/docs/configuration-glossary/configuration-generate/#exclude) those routes when generating and rely on Search Engines to index the JavaScript/SPA version to keep the content current.

The generated sites have the SEO advantages mentioned above plus an easy hosting story. Serving/Hosting static HTML is as fast (and cheap) as it gets.
The user experience is also a plus – browsers consume the fast static HTML on first visit and upon further navigation within the site the SPA kicks in.
